updated README and version (#220)
mario-eth/soldeer: USAGE.md

```bash
[forge] soldeer init
```

#### Soldeer repository

Add the dependencies by using

```bash
[forge] soldeer install <dependency_name>~<version>
```

To search if the dependency is available, visit [https://soldeer.xyz](https://soldeer.xyz).

This command will download the zip file of the dependency, unzip it, install it in the `dependencies` directory.

Additionally, you can install a dependency from any zip file located at any URL:

```bash
[forge] soldeer install <dependency_name>~<version> <url>
```

This command will download the zip file of the dependency from the <url>, unzip it, install it in the `dependencies` directory.

#### Git repositories

The syntax is `soldeer install <dependency>~<version> <git-url>`. This will clone the repository and install the dependency in the `dependencies` folder.

You can also use a certain commit as a dependency

```bash

> soldeer install <dependency>~<version> git:<url> <commit>

Reviewer: ```suggestion
soldeer install <dependency>~<version> <url> <commit>
```

```

Some example

```bash
[forge] soldeer install test-project~v1 git@github.com:test/test.git
[forge] soldeer install test-project~v1 git@gitlab.com:test/test.git
```

```bash
[forge] soldeer install test-project~v1 https://github.com/test/test.git
[forge] soldeer install test-project~v1 https://gitlab.com/test/test.git
```

Or using custom commit hashes

```bash
[forge] soldeer install test-project~v1 git@github.com:test/test.git --rev 345e611cd84bfb4e62c583fa1886c1928bc1a464
```

```bash
[forge] soldeer install
```

> This command will install all the dependencies from the `soldeer.toml`/`foundry.toml` file.

Reviewer: ```suggestion
The command above will install all the dependencies from the `soldeer.toml`/`foundry.toml` file, and will privilege the version inside of the `soldeer.lock` file if present.
```

### How to push a new dependency to the repository

In order to push a new dependency to the repository you have create an account on [https://soldeer.xyz](https://soldeer.xyz), create a project that it will match the dependency name.

Example:

> Create a project called `my-project` and then use the `[forge] soldeer push my-project~v1.0`. This will push the project to the repository and it will be available for everyone to use.

Reviewer: ```suggestion
Create a project called `my-project` and then use the `[forge] soldeer push my-project~1.0.0`. This will push the project to the repository and it will be available for everyone to use.
```

Not sure we should recommend to use a `v` prefix for the version, as that's not the norm in the current packages.

Before using the push command you have to use `[forge] soldeer login` to login to the repository.

#### Pushing a certain directory

If you want to push a certain directory from your project you can use the `[forge] soldeer push my-project~v1.0 /my/path/to/source/files` option. This will push only the files from the specified directory.

#### Ignoring files

> If you want to ignore certain files from the push you need to create a `.soldeerignore` file that will contain the files that you want to ignore. The file should be in the root of the project. This file mimics `.gitignore` syntax.

Reviewer: ```suggestion
If you want to ignore certain files from the push you need to create a `.soldeerignore` file that will contain the files that you want to ignore. The file can be at any level of your directory structure. This file uses the `.gitignore` syntax.
```

#### Dry Run

If you want to dry run a push to inspect what files will be pushed to the central repository, use `[forge] soldeer push my-project~v1.0 [PATH_TO_DEPENDENCY] --dry-run`. This will create a zip file that you can unzip and inspect what was pushed. We recommend everyone to run a dry-run before pushing a new dependency to avoid pushing unwanted files.

**Warning** ⚠️

> You are at risk to push sensitive files to the central repository that then can be seen by everyone. Make sure to exclude sensitive files in the `.soldeerignore` file.

Reviewer: ```suggestion
You are at risk to push sensitive files to the central repository that then can be seen by everyone. Make sure to exclude sensitive files in the `.soldeerignore` or `.gitignore` file.
```

Furthermore, we've implemented a warning that it will be triggered if you try to push a project that contains any `.dot` files/directories.
If you want to skip this warning, you can just use

```bash
[forge] soldeer push my-project~1.0.0 --skip-warnings
```

#### Remappings

The remappings are now fully configurable, the foundry/soldeer TOML files accept a
`[soldeer]` field with the following options

```toml
[soldeer]
# whether soldeer manages remappings

> remappings_generated = true

Reviewer: ```suggestion
remappings_generate = true
```

# whether soldeer re-generates all remappings when installing, updating or uninstalling deps
remappings_regenerate = false

> # whether to suffix the remapping with the version: `name-a.b.c`

Reviewer: ```suggestion
# whether to suffix the remapping with the version requirement string: `name-a.b.c`
```

remappings_version = true

# a prefix to add to the remappings ("@" would give `@name`)
remappings_prefix = ""

# where to store the remappings ("txt" for `remappings.txt` or "config" for `foundry.toml`)
# ignored when `soldeer.toml` is used as config (uses `remappings.txt`)
remappings_location = "txt"

> # whether to install sub-dependencies or not. If true this wil install the dependencies of dependencies 1 level down.

Reviewer: ```suggestion
# whether to install sub-dependencies or not. If true this wil install the dependencies of dependencies recursively.
```

We have a project called `my-project` with the following dependencies:

- `dependency-1`
- `openzeppelin-5.0.2`

A contract inside `my-project` has the following import:

```solidity
@openzeppelin/contracts/token/ERC20/ERC20.sol
```

However, `dependency-1` also requires `openzeppelin`, but it uses version 4.9.2. The contract inside `dependency-1` has the same import:

```solidity
@openzeppelin/contracts/token/ERC20/ERC20.sol
```

Due to improper remappings in the contract files, this situation creates ambiguity, as described above. To resolve this, we should start using versioning within imports, for example:

```solidity
import from 'openzeppelin-4.9.2/token/ERC20/ERC20.sol';
```

This approach will allow us to handle multiple versions of various dependencies effectively.

### Full list of commands

For more commands use `[forge] soldeer --help`.

### CAVEATS

The "add to remappings" feature only appends to the remappings.txt file and does not delete old dependencies. If you want to remove a dependency from remappings, you must do it manually.

> If you use other dependency managers, such as git submodules or npm, ensure you don't duplicate dependencies between soldeer and the other manager.

Reviewer: I think that using the `--regenerate-remappings` option should clean it up no?
